### PR TITLE
Add lru_cache to some hot spots related to sympy manipulation

### PR DESCRIPTION
**Pull request by Chillee**
Reduces compilation time (assuming no autotuning) by about 20-30%.

For ` time  ./benchmarks/torchbench.py -dcuda --inductor --float16 --training --only <model>`

Resnet18: 40s => 30s
densenet121: 2m32s => 2m
hf_Bert: 1m20 => 55s

### CODE REVIEW

File: torchinductor/sizevars.py [hunk 1]
```diff
@@ -348,6 +348,7 @@ def codegen_shape_tuple(self, shape):
         return f"({', '.join(parts)})"
 
 
+@functools.lru_cache(None)
```
Comment (desertfire) [inline torchinductor/sizevars.py:351]: Any other hot functions in this file?

Comment (Chillee) [reply torchinductor/sizevars.py:351]: Not as hot I believe.

We can shave off about another 2-3% by caching on `simplify_sizes_with_ranges`, but it requires a bit more code change due to taking in dicts.

Comment (jansel) [reply torchinductor/sizevars.py:351]: We should have some limit on the cache size other than `None`, perhaps 
```suggestion
@functools.lru_cache(128)
```

Comment (Chillee) [reply torchinductor/sizevars.py:351]: Seems fine to me.

With `cache_size=None`, on densenet:

```
CacheInfo(hits=32555, misses=2486, maxsize=None, currsize=2485)
``

With `cache_size=256`

```
CacheInfo(hits=32023, misses=3072, maxsize=256, currsize=256)
```

**Review by jansel — APPROVED** [review]

**Comment by Chillee** [comment]
I wrapped some more things in `lru_cache` and added a wrapper to `utils.py` to "freeze" inputs.

**Pull request merged**